torch/lib/c10d/ProcessGroupGloo.cpp

```diff
@@ -466,17 +459,20 @@ ProcessGroupGloo::AsyncWork::AsyncWork(
     const char* profilingTitle,
     const c10::optional<std::vector<at::Tensor>>& inputTensors)
     : ProcessGroup::Work(-1, OpType::UNKNOWN, profilingTitle, inputTensors),
-      outputTensors_(std::move(outputTensors)),
-      future_(createFutureAsOutput(outputTensors)) {}
-
-void ProcessGroupGloo::AsyncWork::finishWorkGlooError(std::exception_ptr eptr) {
-  future_->setError(eptr);
-  finish(eptr);
+      outputTensors_(std::move(outputTensors)) {
+  future_->setElementType(futureTypeAsOutput(outputTensors_));
 }
 
-void ProcessGroupGloo::AsyncWork::finishWorkGloo() {
-  returnFutureWithOutput(future_, outputTensors_);
-  finish();
+void ProcessGroupGloo::AsyncWork::finishWithValue() {
+  if (outputTensors_.size() == 0) {
+    finish(c10::IValue(std::vector<at::Tensor>()));
+    return;
+  }
+  if (outputTensors_.size() > 1) {
+    finish(c10::IValue(outputTensors_));
+    return;
+  }
+  finish(c10::IValue(outputTensors_[0]));
```

Comment:
This is unfortunately incorrect when it comes to CUDA tensors, because the H2D copy was performed using the streams in the `streams_` field (in the subclasses that support this) but here we're not setting those streams as current hence the Future will record its events in some arbitrary unrelated streams and thus the Future won't correctly synchronize with the H2D copies.

And I'm realizing only now that `returnFutureWithOutput` seems to suffer from the same problem.